--- Acme-Cow-0.1/cowsay~	Sun May 18 12:42:29 2003
+++ Acme-Cow-0.1/cowsay	Sun May 18 12:46:11 2003
@@ -35,10 +35,11 @@
     'W'		=>	40,
 );
 
-getopts('bde:f:ghlLnNpstT:vwW:y', \%opts);
+getopts('bde:f:ghlLnNprstT:vwW:y', \%opts);
 
 &display_usage() if ($opts{'h'} or $opts{'v'} or $opts{'?'});
 &list_cowfiles() if $opts{'l'};
+&set_random_cow if $opts{'r'};
 
 my $borg = $opts{'b'};
 my $dead = $opts{'d'};
@@ -69,6 +70,34 @@
 $cow->wrap($opts{'W'});
 $cow->print();
 
+sub get_cow_list {
+    my $basedir;
+    my @dirfiles = ();
+    for my $d (split(/$pathsep/, $cowpath)) {
+	opendir(COWDIR, $d) || die "$0: Cannot open $d\n";
+	for my $file (readdir COWDIR) {
+	    if ($file =~ s/\.cow$//) {
+		push(@dirfiles, $file);
+	    }
+	}
+	closedir(COWDIR);
+    }
+    for my $d (@INC) {
+	my @files = glob("$d/Acme/Cow/*.pm");
+	for my $file (@files) {
+	    next if (basename($file) eq "TextBalloon.pm");
+	    next if (basename($file) eq "Example.pm");
+	    push(@dirfiles, basename($file, ".pm"));
+	}
+    }
+    return @dirfiles;
+}
+
+sub set_random_cow {
+    my @cows = &get_cow_list;
+    $opts{'f'} = $cows[int(rand(scalar @cows))];
+}
+
 sub list_cowfiles {
     my $basedir;
     for my $d (split(/$pathsep/, $cowpath)) {
@@ -156,7 +185,7 @@
 	die <<EOF;
 cow{say,think} version $VERSION, (c) 1999-2002 Tony Monroe
 
-Usage: $progname [-bdgpstwy] [-hv?] [-e eyes] [-f cowfile] [-l] [-n] 
+Usage: $progname [-bdgpstwy] [-hrv?] [-e eyes] [-f cowfile] [-l] [-n] 
 	[-T tongue] [-W wrapcolumn] [message]
 
 $rcs_id
@@ -171,7 +200,7 @@
 
 =head1 SYNOPSIS
 
-Usage: $progname [B<-bdgpstwy>] [B<-hv?>] [B<-e> eyes] [B<-f> I<cowfile>] 
+Usage: $progname [B<-bdgpstwy>] [B<-hrv?>] [B<-e> eyes] [B<-f> I<cowfile>] 
           [B<-l>] [B<-n>] [B<-T> I<tongue>] [B<-W> I<wrapcolumn>] [I<message>]
 
 =head1 DESCRIPTION
@@ -222,6 +251,8 @@
 cowfiles installed as modules (in C<@INC>) and on the current
 C<COWPATH>, invoke cowsay with the B<-l> switch.
 
+The B<-r> option makes cowsay select random cowfile (or cowmodule).
+
 If the program is invoked as cowthink (or any word that has "think"
 as a substring) then the cow will think its message instead of
 saying it.
